receiver/vcenterreceiver/scraper.go

```diff
@@ -0,0 +1,373 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vcenterreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/vcenterreceiver"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vsan/types"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/model/pdata"
+	"go.opentelemetry.io/collector/receiver/scrapererror"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/vcenterreceiver/internal/metadata"
+)
+
+// example 2022-03-10 14:15:00
+const timeFormat = "2006-01-02 15:04:05"
+
+var _ component.Receiver = (*vcenterMetricScraper)(nil)
+
+type vcenterMetricScraper struct {
+	client      *vcenterClient
+	mb          *metadata.MetricsBuilder
+	logger      *zap.Logger
+	vsanEnabled bool
+}
+
+func newVmwareVcenterScraper(
+	logger *zap.Logger,
+	config *Config,
+) *vcenterMetricScraper {
+	client := newVmwarevcenterClient(config)
+	return &vcenterMetricScraper{
+		client:      client,
+		logger:      logger,
+		mb:          metadata.NewMetricsBuilder(config.MetricsConfig.Metrics),
+		vsanEnabled: true,
+	}
+}
+
+func (v *vcenterMetricScraper) Start(ctx context.Context, _ component.Host) error {
+	return v.client.Connect(ctx)
+}
+
+func (v *vcenterMetricScraper) Shutdown(ctx context.Context) error {
+	return v.client.Disconnect(ctx)
+}
+
+func (v *vcenterMetricScraper) scrape(ctx context.Context) (pdata.Metrics, error) {
+	if v.client == nil {
+		return pdata.NewMetrics(), errors.New("no SDK client instantiated")
+	}
+
+	// ensure connection before scraping
+	if err := v.client.Connect(ctx); err != nil {
+		return pdata.NewMetrics(), fmt.Errorf("unable to connect to vSphere SDK: %w", err)
+	}
+
+	errs := &scrapererror.ScrapeErrors{}
```

Comment:
This isn't used in this func. Should it be (for example if vsan connection fails)? Otherwise, initialize it where it is used.

Reply:
Not all vCenters have vSAN storage options or the API enabled, so I don't believe it should return any partial error if it cannot reach out to a `/vsanHealth` endpoint if its a normally functioning vCenter just without vSAN. If you feel like we should make a configuration option that controls this more then we can do it that way instead?

Reply:
Ok, let's initialize `errs` where it is used then.